### ISSUE TITLE
 [AC] [Question/Feature]The idea is to optimize the AI action conditions of NPCBOTS and reduce the CPU usage of WorldServer processes.

### ISSUE THREAD

**Issue by moncat2005** [trickerer/Trinity-Bots#706]
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that my issue may get **closed without notice** if I *intentionally* remove or skip any mandatory\* field
- [X] This is NOT a bug/crash report
- [X] I've searched the bugtracker for similar questions/requests **including closed ones**

### Description

After compiling pure AZCORE code, the CPU usage of the worldserver service process is approximately 10% (at this time, there are already 300 NPCBOTS spawned in the world).

After using the npcbots patch code, compile the worldserver program, and the CPU usage during runtime will show around 35%. When there are several players playing online games, it will increase to over 50%.

I guess these CPU consumption may be generated by the autonomous execution of AI actions by npcbots.

When there are no online players, the interaction cost between 300 NPCBOTS objects in the world may be pre spawned. Add the interaction cost between NPCBOTS and players when there are players.

Can we make some optimizations on the AI execution conditions of pre spawn npcbots objects?

For example：
1.NPCBOTS objects with employment relationships can stop executing their AI actions when the hiring player is not online.

2.In addition, unless there is a crazy NPCBOTS refresh, other non employer related NPCBOTs can also consider stop executing AI actions.

I don't know if the above two points can effectively reduce the CPU overhead of WorldServer.

### TrinityCore or AzerothCore

AzerothCore

### Core rev. hash/commit

AzerothCore rev. a9371a0b63b4+ 2023-11-03

### Operating system

Ubuntu 18.04

**Comment by trickerer** [comment]
Stumbling blindly is not the right way to solve a serious problem.

Yes, of course, bot AI update call consumes a considerable amount of CPU, no doubt about that. The question is how much exactly. Most costly checks are only performed in combat - when attacking. Otherwise the amount of CPU cycles consumed by bot AI can be determined via performance profiling.

Actually, all performance issues with idle bots I could find have already been taken care of, where some check would take too long compared to the total CPU time of unit's Update() call. I don't rememeber the exact numbers now but the last time I checked idle bot AI update call was taking less than 10% of total unit's Update() call CPU time.

The main source of slowdown for npcbots the last time I checked was Aura Update() consuming most of unit's Update() cycle, can't elaborate on the details - if passive auras are cheaper in that regard or not for example - but most bots of standard classes have like 20 auras each, even more if the are able to buff each other.

This is the reason number one for bots affecting performance so much. The number two is the fact that any bot in the world keeps entire grid loaded for all times, making all creatures within said grid 'active'. Normally grids not yet visited by any player aren't loaded and creatures within don't exist.
Number three is so called delayed visibity update notifier which is only used by AzerothCore. I may be wrong of course but from my experience this increases average CPU time of visibility updates (which are relatively very expensive) by 70-80%.
And number four, the major one. Default map update interval set by AC devs is 10 milliseconds which means 100(!) Map::Update() ticks per second (including *everything* in loaded grids). For me this tells the worldserver to update the maps as frequenly as it possibly can, unless powerful enough hardware is used, limited only by processor affinity.

Sources 1,2,3 are very difficult to cope with. Note that all of them combine multiplicatively with 4.
I set MapUpdateInterval = 200 in config since my server is kind of small (5 updates a second) and I don't have any kind of performance issues I can't help but notice you have been dealing with for some time now.

The bottom line is: what is your MapUpdateInterval value in config and what happens if you set it to 100 or something (TC default)?

**Comment by moncat2005** [comment]
Thank you for responding to my guess.
In the configuration parameters of my worldserver, map update interval=100, minworldupdatetime=10.

I have tried my best to consider the increase in computational complexity caused by loading other objects into the grid area caused by the spawn point of NPCBOTS.
Originally, I simply selected 300 spawn points for NPCBOTS on the GM Island, but the update time diff value of the world end system increased to over 200 after running continuously for 3 days with up to 10 players online. Players reported significant delays in interface operations.
So I found a remote map location, such as the Hyjal Valley on the blank edge of the world map, and divided it into three gird areas. I distributed the spawn of 300 NPCBOTS according to their professions.
I found that this method can effectively reduce the value of update time diff in the world end system, which is approximately below 30 after one week.

My guess and attempt were just to minimize CPU overhead as much as possible.
This is not because there are many players on my server. In fact, since 2020, there have never been more than 20 real players, just some fans of WOW games that I know.
So, this is just an idealized intention in technology.

Finally, thank you again for the fun your NPCBOTS feature has brought to this game. In a situation where few players are online, the idea of this RPG inspired NPC companion feature is really great.
I will continue to monitor and follow up on your updates, hoping to release compatibility updates for the latest AZ source code.

**Comment by trickerer** [comment]
AC devs still cannot decide on grids unload system and since all grids stay loaded forever you cannot expect a server with any number of players to stay at low CPU usage for a whole week, even without any additional custom code or modules. Even Blizzard servers stay up only for 20 hours.

**Comment by moncat2005** [comment]
Now AZ seems to be in a dilemma with the grid unload system.

I am currently using the method recommended by mpfans: select the AZ version at the end of October and specify several commit numbers related to the grid unload system to be revert (79b39f9, 60e2751, 2779833).

This is not enough, we must revert the commit number (f757e93): "Refactor (Core/Misc): Make DeathState enum class (# 17607)" to address any conflicts with the npcbots feature patch.

I also tried using the az source code from November 15th.
The compilation of only reversing the commit number (f757e93) with npcbots patch is successful.
But I haven't had more players specifically test this version to see if there are any other issues.
On the surface, running for more than a week, occasionally only about 2 players are online, the CPU usage of the WorldServer process is basically around 50%, and the update time diff is also basically normal, below 17.

From this perspective, the main incompatible code of the current npcbots patch seems to be this commit f757e93, which seems to be related to the code rewrite of the DeathState enum class.

I'm not sure if you are already evaluating the compatibility issue with this code and adapting to revise the npcbots patch code?
thanks！

**Comment by trickerer** [comment]
> I'm not sure if you're already evaluating compatibility issues with this code and preparing to modify the npcbots patch code?

https://github.com/trickerer/Trinity-Bots/issues/705#issuecomment-1857456792
Otherwise some enum being made into enum class in not a big problem at all compatibility-wise
